packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts

```diff
@@ -211,3 +172,100 @@ export interface VirtualServiceAttributes {
    */
   readonly clientPolicy?: ClientPolicy;
 }
+
+/**
+ * Properties for a VirtualService provider
+ */
+export interface VirtualServiceProviderConfig {
+  /**
+   * Virtual Node based provider
+   *
+   * @default - none
+   */
+  readonly virtualNodeProvider?: CfnVirtualService.VirtualNodeServiceProviderProperty;
+
+  /**
+   * Virtual Router based provider
+   *
+   * @default - none
+   */
+  readonly virtualRouterProvider?: CfnVirtualService.VirtualRouterServiceProviderProperty;
+
+  /**
+   * Mesh the Provider is using
+   *
+   * @default - none
+   */
+  readonly mesh: IMesh;
+}
+
+/**
+ * Represents the properties needed to define the provider for a VirtualService
+ */
+export abstract class VirtualServiceProvider {
+  /**
+   * Returns a VirtualNode based Provider for a VirtualService
+   */
+  public static virtualNode(virtualNode: IVirtualNode): VirtualServiceProvider {
+    return new VirtualServiceProviderImpl(virtualNode, undefined);
+  }
+
+  /**
+   * Returns a VirtualRouter based Provider for a VirtualService
+   */
+  public static virtualRouter(virtualRouter: IVirtualRouter): VirtualServiceProvider {
+    return new VirtualServiceProviderImpl(undefined, virtualRouter);
+  }
+
+  /**
+   * Returns an Empty Provider for a VirtualService. This provides no routing capabilities
+   * and should only be used as a placeholder
+   */
+  public static none(mesh: IMesh): VirtualServiceProvider {
+    return new VirtualServiceProviderImpl(undefined, undefined, mesh);
+  }
+
+  /**
+   * Mesh the provider belongs to
+   */
+  public abstract readonly mesh: IMesh;
```

Comment:
Let's not expose this (`VirtualServiceProviderConfig` already contains a `readonly mesh: IMesh` property).

Reply:
We need to access this in the Virtual Service Constructor class. The `VirtualServiceProps` only has access to the provider class: `readonly virtualServiceProvider: VirtualServiceProvider;` so we need to access the mesh there to properly create the associated cloudformation.

Reply:
Not sure I understand, sorry.

In the constructor of `VirtualService`, you should call `bind()` on the `virtualServiceProvider` you get, and take the Mesh from there.

There's no point in having 2 separate places where a Mesh is surfaced by.

Reply:
In fact, this is exactly how the code looks today:

```ts
    const provider = props.virtualServiceProvider?.bind(this);
    this.mesh = provider.mesh;
```

So I'm pretty sure this property is unused.

Reply:
https://github.com/aws/aws-cdk/pull/11978/files?file-filters%5B%5D=.ts#diff-c7421c5311bdc768c94ddba1234eb002ff0c1d2d4a5e4e5d31724ea762ec7948R130-R131

Reply:
For some reason I was thinking the bind class was returning an instance of the `VirtualServiceProvider` class 🤦  

Fixed in this revision. Good call